File: androidApp/src/main/java/com/mamba/picme/domain/agent/capability/optimize/gacha/CandidateRenderer.kt
package com.mamba.picme.domain.agent.capability.optimize.gacha

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.mamba.picme.beauty.api.FaceData
import com.mamba.picme.core.common.Logger
import com.mamba.picme.domain.agent.capability.optimize.openImageInputStream
import com.mamba.picme.domain.agent.capability.optimize.recipe.OptimizeRecipeMapper
import com.mamba.picme.features.editor.EditRecipe
import com.mamba.picme.features.editor.RecipeApplier

/**
 * 抽卡候选渲染器：降采样解码 + 经 [RecipeApplier] 渲染候选 preset。
 *
 * 候选一律在 [CANDIDATE_MAX_EDGE] 小图上渲染与评分（速度），
 * 最终应用的全分辨率渲染走编辑器现有路径，不在本类职责内。
 */
class CandidateRenderer(
    private val context: Context,
    private val recipeApplier: RecipeApplier,
    private val faceData: FaceData? = null
) {

    companion object {
        private const val TAG = "PoLang:OptimizeGacha"
        const val CANDIDATE_MAX_EDGE = 512
    }

    /**
     * 解码长边不超过 [maxEdge] 的降采样 Bitmap；失败返回 null（不抛出）。
     * 支持 content://、file:// 与无 scheme 裸路径（chat 附件持久化格式），见 [openImageInputStream]。
     */
    fun decodeDownscaled(imageUri: String, maxEdge: Int = CANDIDATE_MAX_EDGE): Bitmap? {
        return try {
            val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            val boundsStream = openImageInputStream(context, imageUri) ?: run {
                Logger.w(TAG, "decodeDownscaled: openInputStream null: $imageUri")
                return null
            }
            // inJustDecodeBounds 模式下 decodeStream 返回 null 是正常的，只取 outWidth/outHeight
            boundsStream.use { BitmapFactory.decodeStream(it, null, bounds) }
            if (bounds.outWidth <= 0 || bounds.outHeight <= 0) {
                Logger.w(TAG, "decodeDownscaled: bounds decode failed: $imageUri")
                return null
            }

            var sample = 1
            val longEdge = maxOf(bounds.outWidth, bounds.outHeight)
            while (longEdge / (sample * 2) >= maxEdge) sample *= 2

            val options = BitmapFactory.Options().apply { inSampleSize = sample }
            val pixelStream = openImageInputStream(context, imageUri) ?: run {
                Logger.w(TAG, "decodeDownscaled: reopen stream null: $imageUri")
                return null
            }
            pixelStream.use { BitmapFactory.decodeStream(it, null, options) } ?: run {
                Logger.w(TAG, "decodeDownscaled: pixels decode null: $imageUri")
                null
            }
        } catch (e: Exception) {
            Logger.e(TAG, "decodeDownscaled failed: $imageUri", e)
            null
        }
    }

    /** 提取整图像素数组（护栏计算用）。 */
    fun extractPixels(bitmap: Bitmap): IntArray {
        val px = IntArray(bitmap.width * bitmap.height)
        bitmap.getPixels(px, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
        return px
    }

    /**
     * 渲染单个候选 preset。
     *
     * GPU 失败时 [RecipeApplier] 内部已做 CPU 滤镜兜底；此处只捕获未预期异常，
     * 异常返回 null 由编排层丢弃该卡。
     */
    suspend fun render(candidate: OptimizeCandidate, base: Bitmap, sourceUri: String): Bitmap? {
        return try {
            val recipe = OptimizeRecipeMapper.toEditRecipe(
                preset = candidate.preset,
                sourceUri = sourceUri,
                baseRecipe = EditRecipe(sourceUri = sourceUri)
            )
            recipeApplier.applyGpuEffects(base, recipe, faceData)
        } catch (e: Exception) {
            Logger.e(TAG, "render candidate ${candidate.index} (${candidate.direction}) failed", e)
            null
        }
    }
}

File: iosApp/PoLang/Features/Editor/Gacha/CandidateRenderer.swift
import UIKit
import CoreImage
import ImageIO

// MARK: - CandidateRenderer（抽卡候选渲染器）
//
// 移植自 androidApp `domain/agent/capability/optimize/gacha/CandidateRenderer.kt`。
// 降采样解码 + 经 `RecipeApplier` 渲染候选 preset。
//
// 候选一律在 `candidateMaxEdge` 小图上渲染与评分（速度），
// 最终应用的全分辨率渲染走编辑器现有路径，不在本类职责内。
//
// 平台差异（contracts.md C-G5 / editor.yaml §15）：
// - 解码：Android BitmapFactory inSampleSize 两段解码；iOS `CGImageSourceCreateThumbnailAtIndex`
//   （kCGImageSourceThumbnailMaxPixelSize=512）一次降采样，且 withTransform 应用 EXIF 方向
//   （Android decodeStream 不转 EXIF；原图/候选同向自洽，护栏统计不受影响）。
// - 渲染：Android RecipeApplier.applyGpuEffects（自研 GL 引擎，含 beauty 渲染）；
//   iOS RecipeApplier CIImage 管线，beauty 维度因 B1 DEFER 不渲染——人像候选视觉差异仅剩调色，
//   指纹仍含 beauty 维度。滤镜失败兜底在 RecipeApplier 内部（每步退回上一步结果，绝不黑屏）。
final class CandidateRenderer {

    static let candidateMaxEdge = 512
    static let tag = "[PoLang:OptimizeGacha]"

    /// 解码长边不超过 maxEdge 的降采样图；失败返回 nil（不抛出）。
    /// 仅支持本地 file URL（iOS 侧原图统一为 chat/相册落盘文件路径，无 Android content:// 形态）。
    func decodeDownscaled(imageFile: URL,
                          maxEdge: Int = CandidateRenderer.candidateMaxEdge) -> UIImage? {
        guard let src = CGImageSourceCreateWithURL(imageFile as CFURL, nil) else {
            NSLog("%@ decodeDownscaled: CGImageSourceCreateWithURL null: %@", CandidateRenderer.tag, imageFile.path)
            return nil
        }
        let options: [CFString: Any] = [
            kCGImageSourceThumbnailMaxPixelSize: maxEdge,
            kCGImageSourceCreateThumbnailFromImageAlways: true,
            kCGImageSourceCreateThumbnailWithTransform: true,
        ]
        guard let cg = CGImageSourceCreateThumbnailAtIndex(src, 0, options as CFDictionary) else {
            NSLog("%@ decodeDownscaled: thumbnail decode null: %@", CandidateRenderer.tag, imageFile.path)
            return nil
        }
        return UIImage(cgImage: cg)
    }

    /// 提取整图像素（RGBA8 字节缓冲，护栏计算用）。
    ///
    /// 字节序：premultipliedLast = RGBA8 交错（R,G,B,A 每像素 4 字节），与 Android
    /// `Bitmap.getPixels` 的 ARGB IntArray 语义等价（Guardrails 按 RGBA 布局取通道）。
    /// 本管线候选/原图均为不透明照片（JPEG 无 alpha），预乘不影响 RGB 统计。
    func extractPixels(_ image: UIImage) -> [UInt8]? {
        guard let cg = image.cgImage else { return nil }
        let w = cg.width
        let h = cg.height
        guard w > 0, h > 0 else { return nil }
        var buf = [UInt8](repeating: 0, count: w * h * 4)
        let ok = buf.withUnsafeMutableBytes { rawBuffer -> Bool in
            guard let base = rawBuffer.baseAddress,
                  let ctx = CGContext(data: base,
                                      width: w,
                                      height: h,
                                      bitsPerComponent: 8,
                                      bytesPerRow: w * 4,
                                      space: CGColorSpaceCreateDeviceRGB(),
                                      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue)
            else { return false }
            ctx.draw(cg, in: CGRect(x: 0, y: 0, width: w, height: h))
            return true
        }
        return ok ? buf : nil
    }

    /// 渲染单个候选 preset。
    ///
    /// 对齐 Android `CandidateRenderer.render`：候选缩略图一律自全新 `EditRecipe` 起
    /// （不继承用户 crop/markup——对比条只呈现 AI 调色差异）；用户配方的继承发生在
    /// `OptimizeRecipeMapper.toEditRecipe(baseRecipe:)`（AiOptimizeService 落最终配方）。
    /// 滤镜链异常由 RecipeApplier 内部兜底；此处仅捕获未预期失败返回 nil，由编排层丢弃该卡。
    func render(candidate: OptimizeCandidate, base: UIImage, sourceUri: String) -> UIImage? {
        let recipe = OptimizeRecipeMapper.toEditRecipe(
            preset: candidate.preset,
            sourceUri: sourceUri,
            baseRecipe: EditRecipe(sourceUri: sourceUri)
        )
        guard let input = CIImage(image: base) else {
            NSLog("%@ render candidate %d (%@) failed: CIImage conversion null",
                  CandidateRenderer.tag, candidate.index, candidate.direction)
            return nil
        }
        let output = RecipeApplier.apply(recipe, to: input)
        guard let cg = RecipeApplier.context.createCGImage(output, from: output.extent) else {
            NSLog("%@ render candidate %d (%@) failed: createCGImage null",
                  CandidateRenderer.tag, candidate.index, candidate.direction)
            return nil
        }
        return UIImage(cgImage: cg)
    }
}
